src/producer.rs

```diff
@@ -81,6 +81,11 @@ async fn produce_events(ctx: Arc<Context>, start_from: u64, actual: u64) -> eyre
     loop {
         let ctx = ctx.clone();
 
+        if ctx.is_cancelled() {
+            tracing::info!(mode, history_point, "Received cancel signal, stopping...");
+            return Ok(());
+        }
```

Comment:
Looks dumb, but you'll get the idea

```suggestion
        let events_res = select! {
            events_res => ctx.icp().get_events() => { events_res }
            _ => ctx.cancelled() => {
                tracing::info!(mode, history_point, "Received cancel signal, stopping...");
                return Ok(());
            }
        };
        
        le events = events_res.wrap_err_with(|| format!("Failed to get event on history_point: {history_point}"))?;
```